Add expanded description of versioning

noahp opened the pull request.

noahp commented on docs/platform/software-version-hardware-version.mdx:
```diff
@@ -61,7 +61,144 @@ to:
 
-### RTOS
+### Version Schemes
+
```
todo add supported character list:
`-a-zA-Z0-9_.+: ()[]/`

chrisc11 commented on docs/platform/software-version-hardware-version.mdx:
````diff
@@ -61,7 +71,144 @@ to:
+   2. rapid iteration that doesn't require downgrade protection:
+
+      ```plaintext
+      1.2.3+092ced197518c8e304232239315c611af6ff2404
````
For development builds, I've become a fan of the .99 strategy

i.e pre 1.1 release:

```suggestion
      1.0.99
```

or aleternatively

```
1.1.0-dev
```

noahp replied:
I was thinking about the case where they're pushing out eg. nightlies to a cohort- the version identifier would need to change in that case I think? could use incrementing `1.2.3-dev.0` etc, but in case they didn't have CI set up for that... dunno what the best strategy would be :(

noahp replied:
i expanded this section with a few other strategies... could be split off into its own doc, dunno what makes sense!

noahp replied:
also filled it with some fairly opinionated stuff so feel free to push any changes :sweat_smile:

chrisc11 commented on docs/platform/software-version-hardware-version.mdx:
````diff
@@ -61,7 +71,144 @@ to:
+
+   ```plaintext
+   1.2.3
+   2022.04.02
````
might drop this one and just leave the 1.2.3

chrisc11 commented on docs/platform/software-version-hardware-version.mdx:
````diff
@@ -61,7 +71,144 @@ to:
+1. for official releases, strict semver ("normal version number" per the spec):
+
+   ```plaintext
+   1.2.3
````
```suggestion
   1.0.0
```

chrisc11 commented on docs/platform/software-version-hardware-version.mdx:
````diff
@@ -61,7 +71,144 @@ to:
+   2022.10.02
+   ```
+
+   \_Note that mimic the way Memfault's system operates, it may be necessary to
````
```
from natsort import natsorted
natsorted(["1.0.1", "1.0.2"], key=lambda x: x + "z")
['1.0.1', '1.0.2']
```

chrisc11 replied:
```suggestion
   \_Note For exotic versioning schemes, to mimic the way Memfault's system operates, it may be necessary to
```

fnune commented on docs/platform/software-version-hardware-version.mdx:
```diff
@@ -61,7 +71,173 @@ to:
+
+#### Ordering
+
+We recommend using a [SemVer](https://semver.org/) for your software version.
+(See section below)
```
Super nit: actually link to the section, or otherwise improve punctuation

```suggestion
We recommend using a [SemVer](https://semver.org/) for your software version (see section below).
```

fnune commented on docs/platform/software-version-hardware-version.mdx:
```diff
@@ -61,7 +71,173 @@ to:
+devices can tolerate the downgrade).
+
+Memfault uses a slight tweak on top of `natsort` to support `rc` or `dev`
+versioning schemes. This is achieved by appending a character `z` (see
+[here](https://natsort.readthedocs.io/en/master/examples.html#sorting-more-expressive-versioning-schemes)
+for more information).
```
Super nit:
```suggestion
versioning schemes. This is achieved by [appending a character `z`](https://natsort.readthedocs.io/en/master/examples.html#sorting-more-expressive-versioning-schemes).
```
A separate sentence for the link isn't necessary IMO.

fnune commented on docs/platform/software-version-hardware-version.mdx:
````diff
@@ -61,7 +71,173 @@ to:
+```
+
+But natsort will apply lexical ordering (under the default settings), which may
+not result in the expected OTA response (see _"Disable Version Checking"_ above
+for a workaround).
````
Nit: link instead? Less prose, and easier to navigate.

fnune commented on docs/platform/software-version-hardware-version.mdx:
```diff
@@ -61,7 +71,173 @@ to:
+- if using git tags, see information on `git rev-parse` for generating tag
+  strings
+- if using a calendar version, the unix `date` command could be used, for
+  example `date --utc +%Y%m%d`
```
Super nit: missing period on the final bullet point.
```suggestion
  example `date --utc +%Y%m%d`.
```

fnune commented on docs/platform/software-version-hardware-version.mdx:
```diff
@@ -84,20 +260,16 @@ Please see the
-the Software Version is interpreted and sorted, for example using
-[SemVer](https://semver.org/). If you would like to make these changes, please
-let us know.
+By default, irrespective of the source, the Software Version will be ordered per
+the [Version Schemes](#version-schemes) section above.
```
Super nit: the link makes this redundant, and we may move the section.

```suggestion
By default, irrespective of the source, the Software Version will be ordered per
[Version Schemes](#version-schemes).
```

fnune reviewed (approved):
Looks great!

I added a bunch of tiny comments. They're really nitpicky, so feel free to ignore them. It's mostly just me showing my preference.

In the future, I think we should have a style guide for docs, since many of the things I've pointed out here are a matter of taste. But for the sake of consistency...

noahp commented:
> Looks great!
> 
> I added a bunch of tiny comments. They're really nitpicky, so feel free to ignore them. It's mostly just me showing my preference.
> 
> In the future, I think we should have a style guide for docs, since many of the things I've pointed out here are a matter of taste. But for the sake of consistency...

Thank you! these are great, added all of them. Totally agree, it's important to keep the style consistent through our docs.
